webcam-capture-drivers/driver-ffmpeg-cli/src/main/java/com/github/sarxos/webcam/ds/ffmpegcli/FFmpegCliDevice.java

```diff
@@ -1,260 +1,218 @@
+        } catch (IllegalThreadStateException e) {
+            return true;
+        }
+    }
+
+    private Dimension[] readResolutions(String res) {
+        List<Dimension> resolutions = new ArrayList<Dimension>();
+        String[] parts = res.split(" ");
+
+        for (String part : parts) {
+            String[] xy = part.split("x");
+            resolutions.add(new Dimension(Integer.parseInt(xy[0]), Integer.parseInt(xy[1])));
+        }
+
+        return resolutions.toArray(new Dimension[resolutions.size()]);
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public Dimension[] getResolutions() {
+        return resolutions;
+    }
+
+    @Override
+    public Dimension getResolution() {
+        if (resolution == null) {
+            resolution = getResolutions()[0];
+        }
+        return resolution;
+    }
+
+    private String getResolutionString() {
+        Dimension d = getResolution();
+        return String.format("%dx%d", d.width, d.height);
+    }
+
+    @Override
+    public void setResolution(Dimension resolution) {
+        this.resolution = resolution;
+    }
+
+    public void open() {
+        new Thread(new Runnable() {
+            @Override
+            public void run() {
+                open.set(true);
+                try {
+                    startProcess();
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+            }
+        }).start();
+    }
+
+    @Override
+    public void close() {
+        if (!open.compareAndSet(true, false)) {
+            return;
+        }
+
+        process.destroy();
+
+        try {
+            process.waitFor();
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void dispose() {
+        if (disposed.compareAndSet(false, true) && open.get()) {
+            close();
+        }
+    }
+
+    @Override
+    public boolean isOpen() {
+        return open.get();
+    }
+
+    public String[] buildCommand() {
+        return new String[]{
+                "ffmpeg",
+                "-loglevel", "panic",          // suppress ffmpeg headers
+                "-f", "video4linux2",          // camera format
+                "-r", "50",                    // requested FPS
+                "-s", getResolutionString(),   // frame dimension
+                "-i", vfile.getAbsolutePath(), // input file
+                "-vcodec", "rawvideo",         // raw output
+                "-f", "rawvideo",              // raw output
+                "-vsync", "vfr",               // avoid frame duplication
+                "-pix_fmt", "bgr24",           // output format as bgr24
+                "-",                           // output to stdout
+        };
+    }
+
+    public BufferedImage getImage() {
+        if (!open.get()) {
+            return null;
+        }
+        return image;
```

Comment:
This is bad because you are returning the very same instance you are mutating in the other thread. I'm not sure how this will work in case of image non-volatility but I can imagine case when someone is drawing image returned from this method and in the meantime other thread replaced the data inside the image. In this case drawing can be composed of two images - old one and new one.

This should be fixed. You should return immutable data.

```java

BufferedImage bi = new BufferedImage(image.getWidth(), image.getHeight(), BufferedImage.TYPE_INT_RGB);
Graphics2D g2 = bi.createGraphics();

synchronized (image) {
  g2.drawImage(image, 0, 0, null);
}

g2.dispose();

return bi;
```